packer/configfile/configfile.go

```diff
@@ -0,0 +1,47 @@
+package configfile
+
+import (
+	"os"
+	"path/filepath"
+)
+
+const EnvPackerTmpDir = "PACKER_TMP_DIR"
+
+// ConfigFile returns the default path to the configuration file. On
+// Unix-like systems this is the ".packerconfig" file in the home directory.
+// On Windows, this is the "packer.config" file in the application data
+// directory.
+func ConfigFile() (string, error) {
+	return configFile()
+}
+
+// ConfigDir returns the configuration directory for Packer.
+func ConfigDir() (string, error) {
+	return configDir()
+}
+
+// ConfigTmpDir returns a "deterministic" (based on environment or Packer config)
+// path intended as the root of subsequent temporary items, to minimize scatter.
+//
+// The caller must ensure safe tempfile practice via ioutil.TempDir() and friends.
+func ConfigTmpDir() (string, error) {
+	var tmpdir, td string
+	var err error
+
+	if tmpdir = os.Getenv(EnvPackerTmpDir); tmpdir != "" {
+		td, err = filepath.Abs(tmpdir)
+	} else if tmpdir, err = configDir(); err == nil {
+		td = filepath.Join(tmpdir, "tmp")
+	} else if tmpdir = os.TempDir(); tmpdir != "" {
+		td = filepath.Join(tmpdir, "packer")
+	}
```

Comment:
I'm probably out of place in suggesting this since I'm not a packer dever, but I really think a comment that just "suggests" that the caller must ensure safe tempfile practice is not a strong enough enforcement as historically people have produced vulnerable race conditions in this form for nearly ever.. Although, I suppose this logic is to re-implement the way tempfiles are being used so they're not using a potentially shared filesystem anymore and are instead are entirely kept inside a user's home directory though?

Despite packer not needing to be _too_ security conscious, I think it'd be best to have two separate APIs for these two different logics. One api would be something like "ConfigWorkingDirectory" which gives the caller a working directory (similar to PACKER_TMP_DIR but doesn't enforce uniqueness). This way path construction can be hidden away from the caller entirely. (no needing to concatenate manually with configTmpDir)

Then the other would be named something like "TemporaryWorkingDirectory" which would be the same as combining it with `ioutil.TempDir` (which enforces the unique name). In this way it's distinguishable that this can write stuff into a shared directory tree without fear of having more than one instance of packer tamper with files in the same directory.

Reply:
This PR only moved this func here and simplified it a bit ( from `packer/config_file.go`); meaning it was already like this.

The purpose of this pr is to make usage of this func more homogeneous ( or in some cases to use it at all).
That said:
___
Your first suggestion is implemented in the go-getter pull request. It adds a `CachePath` func: https://github.com/hashicorp/packer/pull/6999/files#diff-b841e1ef75f9dff1dde7254a37922dc3 which allows to share downloaded files, ( url+hash is transformed into a hash to confirm content )

Protected with a file lock.
___
On your second suggestion @tb3088 and I agreed to use something like:
```go
tempdir, err := packer.TempDir("hyperv", "path")
// $PACKER_TMP_DIR/hyperv/path
```
Which will give some directories under the same temporary directory; to minimize scatter.

May be we can add a :

```go
tempdir, err := packer.UniqueTempDir("hyperv", "path")
// $PACKER_TMP_DIR/$UNIQUE_PATH/hyperv/path
```

That will do it's best to avoid collisions.

What do you think @tb3088, @arizvisa ?